drivers/adc/adc_stm32.c

```diff
@@ -233,6 +260,96 @@ struct adc_stm32_cfg {
 	size_t pinctrl_len;
 };
 
+#ifdef CONFIG_ADC_STM32_DMA
+static void dma_callback(const struct device *dev, void *arg,
+			 uint32_t channel, int status)
+{
+	/* arg directly holds the adc device */
+	struct adc_stm32_data *data = arg;
+
+	if (channel == data->dma.channel) {
+		if (status != 0) {
+			LOG_ERR("DMA callback error with channel %d.", channel);
+			data->dma_error = status;
+		} else {
+			data->samples_count = data->channel_count;
+		}
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static int adc_stm32_dma_start(const struct device *dev,
+			       const struct adc_sequence *sequence)
+{
+	const struct adc_stm32_cfg *config = dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+	struct adc_stm32_data *data = dev->data;
+	struct dma_block_config *blk_cfg;
+	int ret;
+
+	struct stream *stream = &data->dma;
+
+	blk_cfg = &stream->dma_blk_cfg;
+
+	/* prepare the block */
+	memset(blk_cfg, 0, sizeof(struct dma_block_config));
+	blk_cfg->block_size = sequence->buffer_size;
+
+	/* Source and destination */
+
+	blk_cfg->source_address = (uint32_t)LL_ADC_DMA_GetRegAddr(adc, LL_ADC_DMA_REG_REGULAR_DATA);
+	blk_cfg->source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+	blk_cfg->source_reload_en = 1;
+
+	blk_cfg->dest_address = (uint32_t)sequence->buffer;
+	blk_cfg->dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+	blk_cfg->dest_reload_en = 1;
+
+	/* give the fifo mode from the DT */
+	blk_cfg->fifo_mode_control = data->dma.fifo_threshold;
+
+	/* direction is given by the DT */
+	stream->dma_cfg.head_block = blk_cfg;
+	stream->dma_cfg.user_data = data;
+
+	/* pass our client origin to the dma: data->dma.channel */
+	ret = dma_config(data->dma.dma_dev, data->dma.channel,
+			 &stream->dma_cfg);
+
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		LOG_ERR("Problem setting up DMA %d", ret);
+		return ret;
+	}
+
+	/* invalidate cache to make sure to get the adc data*/
+	SCB_InvalidateDCache_by_Addr(sequence->buffer,
```

Comment:
I think there should be a comment here saying that the buffer needs to be `__aligned(32)`.

Reply:
commented